### DISCUSSION TITLE
Strange behaviour with insert

### DISCUSSION THREAD

**Discussion by revich2** [react-hook-form/react-hook-form#6039] [Q&A]
**Describe the bug**
I have form with fieldArray and elements of fieldArray can add by modal with form for input the name of element. Elements are controlled inputs. In my case, when form with name submitted i need to insert new element in defined place.
When insert happened i see that some elements clones and values from handleSubmit have one more elements than expected.

**To Reproduce**

1. Go to CSB
2. Click button 'Add any element'
3. Type to input and click on 'save'
4. Do 2,3 points again with another value in input
5. You can see that, value which displaying in simple span correct, but value in controlled input isn't correct

**Codesandbox link (Required)**
https://codesandbox.io/s/upbeat-lovelace-2rvmk?file=/src/App.tsx

**Expected behavior**
Insert should be work correct with Controller and second form for this case

**Screenshots**
<img width="424" alt="Screenshot 2021-07-21 at 10 49 36 PM" src="https://user-images.githubusercontent.com/24573190/126550937-4d2b0e5a-559e-4930-a5a3-28c4fb4e7d50.png">

**Desktop (please complete the following information):**

- OS: Big Sur 11.4
- Browser chrome
- Version 91.0.4472.114

**Additional context**
All keys of elements is unique and valid.
I don't know why it happen, but i noticed some regularity. If to do one of this, then all looks like ok:
1) Not use `handleSubmit` in modal form(Simple to fire onSubmit from props)
2) Not use Controller with input

**Comment by bluebill1049** [comment]
react hook form field array is not controlled, if you want to subscribed to changes use `watch` or `useWatch`

**Reply by revich2** [reply]
@bluebill1049 `watch` doesn't help in this case. I tried to add `watch` for field array and behaviour doesn't become right.

CSB - https://codesandbox.io/s/nostalgic-lovelace-yh7tx?file=/src/App.tsx

Watched array contains on one more element then expected.
<img width="552" alt="Screenshot 2021-07-22 at 12 57 33 PM" src="https://user-images.githubusercontent.com/24573190/126621983-6a7f61f5-34bb-4597-801b-211f097caaac.png">

**Reply by revich2** [reply]
@bluebill1049 could you help me with this?

**Reply by bluebill1049** [reply]
can you try this version? react-hook-form@7.12.1-next.0 see if it solve the problem?

**Reply by revich2** [reply]
No, it didn't solve the problem

**Reply by bluebill1049** [reply]
can you simplify your example?

**Reply by revich2** [reply]
i don't know how i can simplify example this example, because it's case from my project.
main of this case are fieldArray and input field with edit mode. i can remove span from render, but i think it doesn't simplify example

**Reply by revich2** [reply]
@bluebill1049 could you make your investigation on this example?

**Reply by bluebill1049** [reply]
<img width="568" alt="Screen Shot 2021-07-27 at 8 21 12 pm" src="https://user-images.githubusercontent.com/10513364/127138385-96b74dd5-6d83-4fdc-a60a-e334386ace40.png">
can you share your steps with a video? i didn't reproduce the issue. 

```tsx
const allFields = fields.map((field, index) => ({
    ...field,
    ...watchedArray[index]
  }));
```

curious why are you doing this above?

**Reply by revich2** [reply]
@bluebill1049 in your screenshot above there is this issue
different values displaying in controller and simple span
<img width="693" alt="Screenshot 2021-07-27 at 8 59 42 PM" src="https://user-images.githubusercontent.com/24573190/127280587-5c5040b0-3570-424b-856d-1801f2c2601f.png">

curious why are you doing this above?

i'm trying to get all possible data from watch and useFieldArray for hope that it can help

**Comment by revich2** [comment]


**Reply by bluebill1049** [reply]
that's because `fields` from useFieldArray only supply defaultValue, if you want to display updated value, use useWatch or watch. back to the circle, hook form is not controlled and does not re-render on each keystroke.

check out this example for more reference: https://github.com/react-hook-form/react-hook-form/blob/master/app/src/useWatchUseFieldArrayNested.tsx

**Reply by bluebill1049** [reply]
> This object contains the defaultValue and key for all your inputs. It's important to assign defaultValue to the inputs.

<img width="970" alt="Screen Shot 2021-07-28 at 9 46 36 am" src="https://user-images.githubusercontent.com/10513364/127241616-9352b141-6620-4f43-a5df-1c3fdd16db3b.png">

**Reply by revich2** [reply]
@bluebill1049 i used `useWatch`, but behaviour the same
CSB - https://codesandbox.io/s/nostalgic-lovelace-yh7tx?file=/src/App.tsx

**Comment by bluebill1049** [comment]
https://codesandbox.io/s/heuristic-rgb-n1flv?file=/src/App.js

**Reply by revich2** [reply]
@bluebill1049 I compared your CSB with my own. And i used my input with edit mode in your CSB and it works fine. After this i saw difference between our examples it is form for adding new item. I used form which created on RHF, you used native html form. And this difference makes this collision.

You can see it in this CSB - https://codesandbox.io/s/sharp-butterfly-xrd05?file=/src/App.js

And videos with different behaviour

Native Form
https://user-images.githubusercontent.com/24573190/127628818-ff6e1237-bed8-4aa4-8e57-f17734d4f7af.mov

Modal with RHF
https://user-images.githubusercontent.com/24573190/127628976-41c00a02-b7a2-447e-a626-2ebd1ca89e70.mov


This behaviour definitely isn't related with watch fields.

And in my first post i describe a couple things which influence on result

>  If to do one of this, then all looks like ok:
> 
> Not use handleSubmit in modal form(Simple to fire onSubmit from props)
> Not use Controller with input

**Reply by revich2** [reply]
@bluebill1049 could you investigate this problem with two RHForms, may be i use these incorrect, but i want use RHF in both cases because i want to use validation and it comfortable

**Reply by revich2** [reply]
@bluebill1049 sorry to disturb, but this case very important for me and i want to solve this problem
could you check it?

**Reply by bluebill1049** [reply]
honestly, I am busy with a lot of stuff, I don't have time to look through your example at the moment. as I have already provided you with examples of how a working example works and spend quite a bit of my time on helping. if there is an issue with this, I suggest you provide a reproducible code sandbox without another dependency. I hope other people in the community can help you out as well.

**Reply by revich2** [reply]
@bluebill1049 i researched this issue and created PR - https://github.com/react-hook-form/react-hook-form/pull/6178
could you check it?

**Reply by revich2** [reply]
@bluebill1049 in PR discussion we found workaround for my case with async `onSubmit` in RHForm - for methods of field array we can use hack like a `reset` API in similar case. But i found case, where this behaviour definitely looks like a bug.

In case when we have a async function, for example api call, and when this promise will fulfilled we can use some method from `useFieldArray` or `reset` and in this case we will get collision like in my case.

I'm sure that this issue can't be solved with some fix in library sources. Of course i was able to fix it with `setTimeout(..., 0)`, but you are right it's dangerous. And i suggest to make a helper that gives to RHForm ability to understand that now some async action is happening.

Something like this:

```jsx
      <form
        onSubmit={(e) => {
          e.preventDefault();
          let formData = new FormData(e.target);

          safeAsyncCall(
            apiCall().then(() => {
              insert(0, {
                name: formData.get("name")
              });
            }),
            { control: form.control }
          );

          e.target.reset();
        }}
      >
```

But i'm not so good in this library sources, may be you know how it can do simplest, or i can do some research and suggest implementation.

I think, that it very important helper, because without it we have form logic which breaks by outside environment.

What do you think about this?

PS:
CSB with this case - https://codesandbox.io/s/stoic-moon-qerx3?file=/src/App.js

**Reply by bluebill1049** [reply]
i will take another look today and give you a definite answer and we can move on from this problem.

**Reply by bluebill1049** [reply]
- mock async: with your PR codesandbox: https://codesandbox.io/s/boring-mccarthy-52lg6?file=/src/App.js looks like its working
- sync: https://codesandbox.io/s/boring-mccarthy-52lg6?file=/src/AddItemForm.js

looks like both working for me. This is the latest mater tho, not published.

**Reply by bluebill1049** [reply]
latest master works with your example as well: https://codesandbox.io/s/kind-rgb-bzjec?file=/src/App.js I think the next release should resolve your issue.

**Reply by revich2** [reply]
@bluebill1049 

> latest master works with your example as well: https://codesandbox.io/s/kind-rgb-bzjec?file=/src/App.js I think the next release should resolve your issue.

i checked this, but problem still exist:
https://user-images.githubusercontent.com/24573190/128607419-1e7ddc10-069e-474f-ac4a-79fc0cb80742.mov

And i checked it on last master branch locally, but here isn't ok

**Reply by bluebill1049** [reply]
works with async now: https://codesandbox.io/s/kind-rgb-bzjec?file=/src/App.js

**Reply by revich2** [reply]
@bluebill1049 really it works, thank you for this fix and test for this case

**Comment by revich2** [comment]
@bluebill1049 Hi, this bug actual again

I found commit with breaking changes - https://github.com/react-hook-form/react-hook-form/commit/dbf6c74b1094adbde7f30645f41d4d981c9cc25d

For comparing you can use this CSB - https://codesandbox.io/s/practical-boyd-tij3l
For version `7.13.0-next.4` it works fine, but `7.13.0-next.5` breaks this behaviour
